### PR TITLE
tree-wide: remove support for kernels lacking ambient caps

### PR DESCRIPTION
**Pull request by poettering**
Let's bump the kernel baseline a bit to 4.3 and thus require ambient caps.

This allows us to remove support for a variety of special casing, most importantly the ExecStart=!! hack.

### CODE REVIEW

**Comment by poettering** [comment]
Let the culling begin.

File: src/core/load-fragment.c [hunk 1]
```diff
@@ -944,28 +944,27 @@ int config_parse_exec(
...
+                        else if (*f == '!' && !(flags & (EXEC_COMMAND_FULLY_PRIVILEGED) && !ambient_hack)) {
+                                /* Compatibility with the old !! ambient caps hack (removed in v258). Since
+                                 * we don't support that anymore and !! was a noop on non-supporting systems,
+                                 * we'll just turn off the EXEC_COMMAND_NO_SETUID flag again and be done with
+                                 * it. */
                                 flags &= ~EXEC_COMMAND_NO_SETUID;
-                                flags |= EXEC_COMMAND_AMBIENT_MAGIC;
+                                ambient_hack = true;
+
+                                log_syntax(unit, LOG_NOTICE, filename, line, 0, "!! modifier for %s= lines is no longer supported and is now ignored. Please update your unit files and remove the modifier.", lvalue);
```
Comment (bluca) [inline src/core/load-fragment.c]: LOG_WARNING maybe?

File: man/systemd.exec.xml [hunk 2]
```diff
@@ -832,10 +832,11 @@ CapabilityBoundingSet=~CAP_B CAP_C</programlisting>
         assigned to this option, the ambient capability set is reset to the empty capability set, and all prior
         settings have no effect. If set to <literal>~</literal> (without any further argument), the ambient capability
         set is reset to the full set of available capabilities, also undoing any previous settings. Note that adding
-        capabilities to the ambient capability set adds them to the process's inherited capability set. </para><para>
-        Ambient capability sets are useful if you want to execute a process as a non-privileged user but still want to
-        give it some capabilities. Note that in this case option <constant>keep-caps</constant> is automatically added
-        to <varname>SecureBits=</varname> to retain the capabilities over the user
+        capabilities to the ambient capability set adds them to the process's inherited capability set. </para>
```
Comment (YHNdnzj) [inline man/systemd.exec.xml]:
```suggestion
        capabilities to the ambient capability set adds them to the process's inherited capability set.</para>
```

File: src/shared/exec-util.h [hunk 3]
```diff
@@ -48,9 +48,9 @@ typedef enum ExecCommandFlags {
         EXEC_COMMAND_IGNORE_FAILURE   = 1 << 0,
         EXEC_COMMAND_FULLY_PRIVILEGED = 1 << 1,
         EXEC_COMMAND_NO_SETUID        = 1 << 2,
-        EXEC_COMMAND_AMBIENT_MAGIC    = 1 << 3,
-        EXEC_COMMAND_NO_ENV_EXPAND    = 1 << 4,
+        EXEC_COMMAND_NO_ENV_EXPAND    = 1 << 3,
         _EXEC_COMMAND_FLAGS_INVALID   = -EINVAL,
+        _EXEC_COMMAND_FLAGS_MAX       = (1 << 4) -1,
```
Comment (YHNdnzj) [inline src/shared/exec-util.h]: Hmm, is this really useful? And at least this should be `_ALL`, not `_MAX`.

**Review by YHNdnzj — APPROVED** [review]

File: src/shared/exec-util.h (hunk 3 above)
Comment (poettering) [reply src/shared/exec-util.h]: yes, it is useful. I added an assert_cc() to exec-util.c that makes use of this and ensures the associated string array is complete.

File: src/core/load-fragment.c (hunk 1 above)
Comment (poettering) [reply src/core/load-fragment.c]: In almost all cases where we log about an obsolete option that we now ignore we use LOG_NOTICE, hence we should do so here too.

And I think that makes sense: a warning is about some error we gracefully handle, but this here is an option we now ignore which is really far from an error: the fact that we ignore it now changes exactly nothing in the behaviour, it didn't have any effect on any recent systems anyway. Hence I think LOG_WARNING is too much (but LOG_INFO would be too little), hence LOG_NOTICE it is, like in the other cases.

Comment (bluca) [reply src/core/load-fragment.c]: Ok

File: src/core/load-fragment.c [hunk 4]
```diff
@@ -944,28 +944,27 @@ int config_parse_exec(
...
                         else if (*f == ':' && !(flags & EXEC_COMMAND_NO_ENV_EXPAND))
                                 flags |= EXEC_COMMAND_NO_ENV_EXPAND;
-                        else if (*f == '+' && !(flags & (EXEC_COMMAND_FULLY_PRIVILEGED|EXEC_COMMAND_NO_SETUID|EXEC_COMMAND_AMBIENT_MAGIC)))
+                        else if (*f == '+' && !(flags & (EXEC_COMMAND_FULLY_PRIVILEGED|EXEC_COMMAND_NO_SETUID)) && !ambient_hack)
                                 flags |= EXEC_COMMAND_FULLY_PRIVILEGED;
-                        else if (*f == '!' && !(flags & (EXEC_COMMAND_FULLY_PRIVILEGED|EXEC_COMMAND_NO_SETUID|EXEC_COMMAND_AMBIENT_MAGIC)))
+                        else if (*f == '!' && !(flags & (EXEC_COMMAND_FULLY_PRIVILEGED|EXEC_COMMAND_NO_SETUID)) && !ambient_hack)
                                 flags |= EXEC_COMMAND_NO_SETUID;
-                        else if (*f == '!' && !(flags & (EXEC_COMMAND_FULLY_PRIVILEGED|EXEC_COMMAND_AMBIENT_MAGIC))) {
+                        else if (*f == '!' && !(flags & (EXEC_COMMAND_FULLY_PRIVILEGED) && !ambient_hack)) {
```
Comment (yuwata) [inline src/core/load-fragment.c]: Spurious position of `)` ??
```c
else if (*f == '!' && !(flags & (EXEC_COMMAND_FULLY_PRIVILEGED)) && !ambient_hack) {
```

File: src/core/load-fragment.c (hunk 1 above)
Comment (yuwata) [inline src/core/load-fragment.c]: Not a matter, the line is slightly long. I'd like to break the line after `0,`.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/core/load-fragment.c (hunk 4 above)
Comment (YHNdnzj) [reply src/core/load-fragment.c]: Ugh, should probably just use FLAGS_SET

**Review by yuwata — APPROVED** [review]

**Pull request merged**